Sources/Apodini/SyntaxTreeVisitor/ContextNode.swift

```diff
@@ -28,31 +29,43 @@ class ContextNode {
             ?? getGlobalContextValue(for: contextKey)
             ?? C.defaultValue
     }
+
+    func getContextValue<C: OptionalContextKey>(for contextKey: C.Type = C.self) -> C.Value? {
+        getNodeOnlyContextValue(for: contextKey)
+            ?? getGlobalContextValue(for: contextKey)
+    }
     
-    private func getNodeOnlyContextValue<C: ContextKey>(for contextKey: C.Type = C.self) -> C.Value? {
+    private func getNodeOnlyContextValue<C: OptionalContextKey>(for contextKey: C.Type = C.self) -> C.Value? {
         nodeOnlyContext[ObjectIdentifier(contextKey)] as? C.Value
     }
     
-    private func getNodeContextValue<C: ContextKey>(for contextKey: C.Type = C.self) -> C.Value? {
+    private func getNodeContextValue<C: OptionalContextKey>(for contextKey: C.Type = C.self) -> C.Value? {
         context[ObjectIdentifier(contextKey)] as? C.Value
     }
     
-    private func getGlobalContextValue<C: ContextKey>(for contextKey: C.Type = C.self) -> C.Value? {
+    private func getGlobalContextValue<C: OptionalContextKey>(for contextKey: C.Type = C.self) -> C.Value? {
         getNodeContextValue(for: contextKey) ?? parentContextNode?.getGlobalContextValue(for: contextKey)
     }
     
-    func addContext<C: ContextKey>(_ contextKey: C.Type = C.self, value: C.Value, scope: Scope) {
-        var newValue: C.Value
+    func addContext<C: OptionalContextKey>(_ contextKey: C.Type = C.self, value: C.Value, scope: Scope) {
+        let newValue: C.Value
+
+        if C.Value.self is ExpressibleByNilLiteral.Type {
```

Comment:
I am not sure whether this check is really the best type to determine if `Value` is an optional type.
There might be cases where custom types conform to `ExpressibleByNilLiteral`.

I search-engined a bit, and found that [`Mirror.DisplayStyle`](https://developer.apple.com/documentation/swift/mirror/displaystyle) has an `optional` case. 
Maybe that's a nicer approach, even though still only a runtime-check 🤔 :

```swift
func isOptional(_ instance: Any) -> Bool {
    let mirror = Mirror(reflecting: instance)
    return mirror.displayStyle == .optional
}
```
(stolen from https://stackoverflow.com/a/48690919)

Reply:
> There might be cases where custom types conform to ExpressibleByNilLiteral.

Our stance is, we don't allow `nil` literal, and thus I think the check for that protocol is actually valid(?). We do not just blacklist `Optional` type, but every type which allows `nil`(?).

In any case, this is internal API, so not sure if that is that big of a problem. But I assume adding a private protocol and conformance for that protocol to `Optional` would also be valid.

Reply:
@Supereg @moritzschuell The `Runtime` framework offers a build in check to detect an optional type, we should be able to use this here. I have experimented a bit with this and created a test case to evaluate the function and it seems to work in my limits tests:
```swift
func testIsOptional() throws {
    func isOptional<T>(_ type: T.Type = T.self) throws -> Bool {
        let typeInfo = try Runtime.typeInfo(of: type)
        return typeInfo.kind == .optional
    }
    
    /// A custom type
    struct Test {}
    
    XCTAssertEqual(try isOptional(String.self), false)
    XCTAssertEqual(try isOptional(Int.self), false)
    XCTAssertEqual(try isOptional(Test.self), false)
    XCTAssertEqual(try isOptional(Optional<Test>.self), true)
    XCTAssertEqual(try isOptional(Optional<Test>.self), true)
    XCTAssertEqual(try isOptional(String?.self), true)
    XCTAssertEqual(try isOptional(String??.self), true)
    XCTAssertEqual(try isOptional(String???.self), true)
    XCTAssertEqual(try isOptional(Never.self), false)
}
```

Therefore we could introduce an internal method similar to `assertTypeIsStruct` that can look like this: 
```swift
func isOptional<T>(_ type: T.Type = T.self) throws -> Bool {
    let typeInfo = try Runtime.typeInfo(of: type)
    return typeInfo.kind == .optional
}
```
and could be used here:
```suggestion
        if isOptional(C.Value.self) {
```

Reply:
I incorporated your proposed change @PSchmiedmayer with 717482c.
I just modified the isOptional func to not be throwing (with documentation why this is possible).
Basically `typeInfo` only throws if the `kind` is something it does not support (or is not a "type"). If the kind is optional it will never throw, as it will generate a enum TypeInfo. Thus if typeInfo throws, we know that it can't be an optional.
I think that makes using the function a bit easier.

I also just added the test cases you provided 👍

Reply:
Thats a good point, making it non throwing is a good addition 👍